src/main/java/io/cdap/plugin/gcp/publisher/source/PubSubSubscriber.java

```diff
@@ -0,0 +1,183 @@
+import io.cdap.cdap.api.annotation.Macro;
+import io.cdap.cdap.api.data.schema.Schema;
+import io.cdap.cdap.etl.api.FailureCollector;
+import io.cdap.cdap.etl.api.PipelineConfigurer;
+import io.cdap.cdap.etl.api.streaming.StreamingContext;
+import io.cdap.cdap.etl.api.streaming.StreamingSource;
+import io.cdap.cdap.etl.api.streaming.StreamingSourceContext;
+import io.cdap.plugin.common.LineageRecorder;
+import io.cdap.plugin.gcp.common.GCPReferenceSourceConfig;
+import io.cdap.plugin.gcp.common.GCPUtils;
+import org.apache.spark.SparkConf;
+import org.apache.spark.storage.StorageLevel;
+import org.apache.spark.streaming.api.java.JavaDStream;
+import org.apache.spark.streaming.api.java.JavaReceiverInputDStream;
+import org.apache.spark.streaming.dstream.ReceiverInputDStream;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import scala.reflect.ClassTag;
+
+import java.util.stream.Collectors;
+import javax.annotation.Nullable;
+
+/**
+ * Base implementation of a Realtime source plugin to read from Google PubSub.
+ *
+ * @param <T> The type that the PubSubMessage will be mapped to by the mapping function.
+ */
+public abstract class PubSubSubscriber<T> extends StreamingSource<T> {
+
+  protected static final Logger LOG = LoggerFactory.getLogger(PubSubSubscriber.class);
+
+  protected SubscriberConfig config;
+  protected Schema schema;
+  protected SerializableFunction<PubSubMessage, T> mappingFunction;
+
+  public PubSubSubscriber(SubscriberConfig conf, Schema schema,
+                          SerializableFunction<PubSubMessage, T> mappingFunction) {
+    this.config = conf;
+    this.schema = schema;
+    this.mappingFunction = mappingFunction;
+  }
+
+  @Override
+  public void configurePipeline(PipelineConfigurer pipelineConfigurer) {
+    super.configurePipeline(pipelineConfigurer);
+    pipelineConfigurer.getStageConfigurer().setOutputSchema(this.schema);
+  }
+
+  @Override
+  public void prepareRun(StreamingSourceContext context) throws Exception {
+    FailureCollector collector = context.getFailureCollector();
+    config.validate(collector);
+    // record dataset lineage
+    context.registerLineage(config.referenceName, this.schema);
+
+    if (this.schema.getFields() != null) {
+      LineageRecorder recorder = new LineageRecorder(context, config.referenceName);
+      recorder.recordRead("Read", "Read from Pub/Sub",
+                          this.schema.getFields().stream().map(Schema.Field::getName).collect(Collectors.toList()));
+    }
+  }
+
+  @Override
+  public JavaDStream<T> getStream(StreamingContext streamingContext) throws Exception {
+    SparkConf sparkConf = streamingContext.getSparkStreamingContext().ssc().conf();
+    sparkConf.set("spark.streaming.receiver.writeAheadLog.enable", "true");
+    sparkConf.set("spark.streaming.driver.writeAheadLog.closeFileAfterWrite", "true");
+    sparkConf.set("spark.streaming.receiver.writeAheadLog.closeFileAfterWrite", "true");
```

Comment:
These two settings are checkpoint fs specific. E.g. if the checkpoint is on HDFS, you actually don't want these set to true.

Reply:
I've removed these properties.